br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -573,6 +590,127 @@ func (m *DupeDetector) RecordIndexConflictError(ctx context.Context, stream DupK
+// according to key and value.
+func NewErrFoundConflictRecords(key []byte, value []byte, tbl table.Table) error {
+	sessionOpts := encode.SessionOptions{
+		SQLMode: mysql.ModeStrictAllTables,
+	}
+
+	decoder, err := kv.NewTableKVDecoder(tbl, tbl.Meta().Name.L, &sessionOpts, log.L())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if tablecodec.IsRecordKey(key) {
+		// for data KV
+		handle, err := tablecodec.DecodeRowKey(key)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		rowData := decoder.DecodeRawRowDataAsStr(handle, value)
+
+		return errors.Trace(common.ErrFoundDataConflictRecords.FastGenByArgs(tbl.Meta().Name, handle.String(), rowData))
+	}
+
+	// for index KV
+	_, idxID, _, err := tablecodec.DecodeIndexKey(key)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	idxInfo := model.FindIndexInfoByID(tbl.Meta().Indices, idxID)
+	return NewErrFoundIndexConflictRecords(key, value, tbl, idxInfo)
+}
+
+// GenIndexValueFromIndex generate index value from index.
+func GenIndexValueFromIndex(key []byte, value []byte, tblInfo *model.TableInfo, idxInfo *model.IndexInfo) ([]string, error) {
+	idxColLen := len(idxInfo.Columns)
+	colInfos := tables.BuildRowcodecColInfoForIndexColumns(idxInfo, tblInfo)
+	values, err := tablecodec.DecodeIndexKV(key, value, idxColLen, tablecodec.HandleNotNeeded, colInfos)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	valueStr := make([]string, 0, idxColLen)
+	for i, val := range values[:idxColLen] {
+		d, err := tablecodec.DecodeColumnValue(val, colInfos[i].Ft, time.Local)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		str, err := d.ToString()
+		if err != nil {
+			str = string(val)
+		}
+		if types.IsBinaryStr(colInfos[i].Ft) || types.IsTypeBit(colInfos[i].Ft) {
+			str = tidbutil.FmtNonASCIIPrintableCharToHex(str)
+		}
+		valueStr = append(valueStr, str)
+	}
+
+	return valueStr, nil
+}
+
+// NewErrFoundIndexConflictRecords generate an error ErrFoundIndexConflictRecords
+// according to key and value.
+func NewErrFoundIndexConflictRecords(key []byte, value []byte, tbl table.Table, idxInfo *model.IndexInfo) error {
+	sessionOpts := encode.SessionOptions{
+		SQLMode: mysql.ModeStrictAllTables,
+	}
+
+	decoder, err := kv.NewTableKVDecoder(tbl, tbl.Meta().Name.L, &sessionOpts, log.L())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	h, err := decoder.DecodeHandleFromIndex(idxInfo, key, value)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	valueStr, err := GenIndexValueFromIndex(key, value, tbl.Meta(), idxInfo)
+	indexName := fmt.Sprintf("%s.%s", tbl.Meta().Name.String(), idxInfo.Name.String())
+	if err != nil {
+		log.L().Warn("decode index key value / column value failed", zap.String("index", indexName),
+			zap.String("key", hex.EncodeToString(key)), zap.String("value", hex.EncodeToString(value)), zap.Error(err))
+		return errors.Trace(common.ErrFoundIndexConflictRecords.FastGenByArgs(tbl.Meta().Name, indexName, key, value))
```

Comment:
in this case, `key` and `value` does noy satisfy the meaning of message template of ErrFoundIndexConflictRecords. Should we use a new error?

Reply:
Here I conform to the original `genKeyExistsErr` function, where it also returns the expected `ErrKeyExists` error with the raw key
https://github.com/pingcap/tidb/blob/897d93db8c9b0959803cc7ebae419bf72da0ded3/pkg/ddl/index.go#L1604
I think here we can also use the raw key and value, they satisfy the meaning of ErrFoundIndexConflictRecords, although they are in the form of raw but not human readable.

Reply:
The error message template of `ErrKeyExists` is "Duplicate entry '%-.64s' for key '%-.192s'". We have a different template. Arguments order and meaning is wrong, not whether it's raw or not.